server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader10to11.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.accumulo.server.conf.store.TablePropKey;
+import org.apache.hadoop.fs.Path;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class Upgrader10to11 implements Upgrader {
+
+  private static final Logger log = LoggerFactory.getLogger(Upgrader10to11.class);
+
+  // Included for upgrade code usage any other usage post 3.0 should not be used.
+  private static final TableId REPLICATION_ID = TableId.of("+rep");
+
+  public Upgrader10to11() {
+    super();
+  }
+
+  @Override
+  public void upgradeZookeeper(final ServerContext context) {
+    log.info("upgrade of ZooKeeper entries");
+
+    var zrw = context.getZooReaderWriter();
+    var iid = context.getInstanceID();
+
+    // if the replication base path (../tables/+rep) assume removed or never existed.
+    if (!checkReplicationTableInZk(iid, zrw)) {
+      log.debug("replication table root node does not exist in ZooKeeper - nothing to do");
+      return;
+    }
+
+    // if the replication table is online - stop. There could be data in transit.
+    if (!checkReplicationOffline(iid, zrw)) {
+      throw new IllegalStateException(
+          "Replication table is not offline. Cannot continue with upgrade that will remove replication with replication active");
+    }
+
+    deleteReplicationConfigs(zrw, iid, context.getPropStore());
+
+    deleteReplicationTableZkEntries(zrw, iid);
+
+  }
+
+  @Override
+  public void upgradeRoot(final ServerContext context) {
+    log.info("upgrade root - skipping, nothing to do");
+  }
+
+  @Override
+  public void upgradeMetadata(final ServerContext context) {
+    log.info("upgrade metadata entries");
+    deleteReplMetadataEntries(context);
+    deleteReplHdfsFiles(context);
+  }
+
+  /**
+   * remove +rep entries from metadata.
+   */
+  private void deleteReplMetadataEntries(final ServerContext context) {
+    try (BatchDeleter deleter =
+        context.createBatchDeleter(MetadataTable.NAME, Authorizations.EMPTY, 10)) {
+
+      Range repTableRange =
+          new Range(REPLICATION_ID.canonical() + ";", true, REPLICATION_ID.canonical() + "<", true);
+      // copied from MetadataSchema 2.1 (removed in 3.0)
+      Range repWalRange =
+          new Range(RESERVED_PREFIX + "repl", true, RESERVED_PREFIX + "repm", false);
+
+      deleter.setRanges(List.of(repTableRange, repWalRange));
+      deleter.delete();
+    } catch (TableNotFoundException | MutationsRejectedException ex) {
+      throw new IllegalStateException("failed to remove replication info from metadata table", ex);
+    }
+  }
+
+  @VisibleForTesting
+  void deleteReplHdfsFiles(final ServerContext context) {
+    try {
+      for (Volume volume : context.getVolumeManager().getVolumes()) {
+        String dirUri = volume.getBasePath() + Constants.HDFS_TABLES_DIR + Path.SEPARATOR
+            + REPLICATION_ID.canonical();
```

Comment:
This is making assumptions that there are no other files in these directories other than files used by the replication table and not used by any other table. That doesn't seem like a safe assumption. For example, if the replication table had been cloned, some of its files might now be referenced by the clone.

Instead of deleting these directories, we should instead read the file entriess in the replication table's tablets and create corresponding `~del` entries for the `accumulo-gc` to delete later when it determines they no longer need to be preserved. This also allows the actual deletion to be deferred by the user by avoiding running the `accumulo-gc` during upgrade (which might be one way they could upgrade more cautiously) or by utilizing the HDFS trash that the `accumulo-gc` is configured to use.

Reply:
Will need to look at this - during the upgrade, and depending on the phase, there is limited information - what is in zookeeper, what is in the root table and what is in the metadata - other tables have not been brought online. 

One way to handle this could be through instructions.

The replication table must be offline for the upgrade to proceed.  So, if someone is using replication, the instructions could include the warning that any tables that may have been cloned / shared data with the replication table, they should be compacted prior to upgrading so they do not share any files.

Alternatively, the files could be left in place and leave it up to the user to delete the hdfs directory if they want.  

Removing the zookeeper and metadata entries during the upgrade process seems to be required because of the "special" status as a system table, but also closes off some other options.

Reply:
Addressed with 8490bde61d - reads the metadata table for replication table files and then deletes the files individually.  Un-referenced files and the dir structure is left alone.